Renaming ZooStore get/set property methods to get/set transactionInfo

cshannon opened the pull request:
Also using a new enum type for the txInfo names instead of String values.

This closes #2841

cshannon commented:
@EdColeman - Here is my first attempt at renaming the methods and using an enum. Per our discussion offline I renamed the "debug" property to "target" as it's a better description of what it is. Obviously any of those names can be changed or tweaked based on what people think. I didn't add any new tests as I figure the existing Fate tests should catch any issues with the ZooStore and refactoring.

EdColeman reviewed (commented):
There is a subclass AdminUtil.TransactionStatus that has `debug` as a field

EdColeman commented on core/src/main/java/org/apache/accumulo/core/logging/FateLogger.java:
```diff
@@ -80,8 +81,8 @@ public TStatus waitForStatusChange(long tid, EnumSet<TStatus> expected) {
       @Override
-      public Serializable getProperty(long tid, String prop) {
-        return store.getProperty(tid, prop);
+      public Serializable getNodeData(long tid, Fate.NodeData prop) {
```
Or `nodeData` as used elsewhere

dlmarion replied:
`getTransactionData`, `getTransactionInfo`? See my comment for `Fate.NodeData`.

cshannon commented:
@EdColeman - Thanks for the quick review, I'll look at making these changes when I get some time later this week either tomorrow or this weekend.

dlmarion commented on core/src/main/java/org/apache/accumulo/fate/Fate.java:
```diff
@@ -70,6 +65,29 @@
 
   private final AtomicBoolean keepRunning = new AtomicBoolean(true);
 
+  public enum NodeData {
```
This enum contains members whose names correspond to the attributes of an instance of a Fate transaction. When storing or accessing these attributes you have to supply the Fate transaction id. I would be more inclined to support a name like `TransactionData`, `TxData`, or something like that. I think the word "node" is coming from the fact that the data is stored in a ZK node. The name isn't really descriptive to someone looking at this code that doesn't know the inner workings of the persistence.

EdColeman replied:
`TxData` is nice and short and provides context and may be an improvement over node

dlmarion commented on core/src/main/java/org/apache/accumulo/fate/Fate.java:
```diff
@@ -70,6 +65,29 @@
 
+  public enum NodeData {
+
+    TARGET_NODE("target_node"),
```
I think we can remove `_node` from all of the enum names. Also, I would support renaming `Repo.getDescription` to `Repo.getName` or `Repo.getClassName` as the class name is only ever used. Then, instead of using `target` here, I would just use `name`.

EdColeman replied:
Without `_node` the exception and return nodes seem like the should have some decoration.  The `target` (formally prop_debug) is the classname of the REPO that is the "head" of the transaction (the "tail" of the stack) - the REPO that starts things, but can have sub-steps.   The distinction of that "name" should be reflected in the code.  Things like repoHeadName seem awkward.  To contrast with top that is the currently executing transaction, tail - but then that requires a deep understanding of the stack nature of REPO.

dlmarion replied:
If you take my comment about `Fate.NodeData` into account and rename the enum, then you would end up with something like:
```
enum TxInfo {
  NAME, EXCEPTION, RETURN, AUTO_CLEAN;
}
```

> Without _node the exception and return nodes seem like the should have some decoration.

I think `TxInfo.EXCEPTION` is pretty clear. Maybe `TxInfo.RETURN_VALUE` is a little more clear.

EdColeman replied:
Both `TxInfo.EXCEPTION` and `TxInfo.RETURN_VALUE` seem good.  `NAME` feels to generic - what about `TxInfo.REPO_NAME`

dlmarion replied:
> what about TxInfo.REPO_NAME

Sure, IMO better than `target` and `debug`. I would still suggest renaming `Repo.getDescription`, maybe in this case to `Repo.getName`.

dlmarion replied:
@cshannon - @EdColeman and I talked offline and we settled on the following structure for the enum:
```
TxInfo {
  REPO_TARGET, EXCEPTION, RETURN_VALUE, AUTO_CLEAN
}
```

milleruntime replied:
Did you guys mean for the first value in the Enum to be "REPO_NAME" instead? While working on the new Fate API, I came to parts of the code using the field and I found that the term "target" doesn't make sense. https://github.com/apache/accumulo/blob/e08287065d3af4773dd97a3991686abc3002f6dd/core/src/main/java/org/apache/accumulo/fate/Fate.java#L303

milleruntime replied:
The description [here](https://github.com/apache/accumulo/blob/e08287065d3af4773dd97a3991686abc3002f6dd/core/src/main/java/org/apache/accumulo/fate/AdminUtil.java#L108) made me scratch my head. When we print out the value on the CLI, its just referred to as "op", meaning the name of the current operation. https://github.com/apache/accumulo/blob/e08287065d3af4773dd97a3991686abc3002f6dd/core/src/main/java/org/apache/accumulo/fate/AdminUtil.java#L421-L425

Here is a Fate Tx printed for a compaction:
<pre>
txid: 2d4b1f3195963320  status: IN_PROGRESS         op: CompactRange     locked: [R:+default, R:1] locking: []              top: CompactionDriver created: 2022-08-30T19:30:00.06Z
</pre>

So "CompactRange" is really just the name of the current operation running for the Tx.

dlmarion replied:
In the comment thread above, `name` and `repoName` were both suggested. I don't remember how we ended up at repoTarget.

milleruntime replied:
I made a fix in #2904

cshannon commented:
@EdColeman and @dlmarion - I updated my PR with a new commit to address the comments. I renamed the enum to TxInfo and the methods. Also another change is `the` enum name is used as is now when storing data into ZK which is all caps instead of lower case (this can be changed of course but seems ok to me).

dlmarion reviewed (approved):
LGTM

milleruntime commented on core/src/main/java/org/apache/accumulo/fate/ReadOnlyTStore.java:
```diff
@@ -129,10 +129,10 @@ enum TStatus {
-   * @param prop
-   *          name of property to retrieve.
+   * @param txInfo
+   *          name of attribute of a transaction to retrieve.
```
Javadoc above still refers to "property"

milleruntime commented on core/src/main/java/org/apache/accumulo/fate/Fate.java:
```diff
@@ -71,6 +66,10 @@
   private final AtomicBoolean keepRunning = new AtomicBoolean(true);
 
+  public enum TxInfo {
+    REPO_TARGET, AUTO_CLEAN, EXCEPTION, RETURN_VALUE;
```
```suggestion
    REPO_TARGET, AUTO_CLEAN, EXCEPTION, RETURN_VALUE
```

milleruntime commented on core/src/main/java/org/apache/accumulo/fate/TStore.java:
```diff
@@ -65,7 +65,17 @@
 
-  void setProperty(long tid, String prop, Serializable val);
+  /**
+   * Set a transaction-specific property.
```
```suggestion
   * Set transaction-specific information.
```

milleruntime reviewed (commented):
Looks good but there were a few references to old name (property) still hanging around in the javadoc.

cshannon commented:
> Looks good but there were a few references to old name (property) still hanging around in the javadoc.

Thanks, I can fix those tomorrow morning and then we can hopefully get this merged in.

cshannon commented:
@milleruntime - Ok I added a new commit to fix the Javadoc and enum. This should be ready to merged into main I think.

dlmarion commented:
I'm going to kick off a Full IT build as a sanity check

dlmarion commented:
Full IT build passed
